### PR TITLE
CLI refactor

### LINKED ISSUE #369: CLI default config
- add to $HOME/.kic and $HOME/.flt
- default to $HOME/bin/.kic
- remove references to $KIC_PATH env var

### LINKED ISSUE #421: Integrate Coral and PiB
- the spike was very successful
- this will be a breaking change and fleets will have to be rebuilt
- this gives PiB
  - workspaces
  - kustomization per workspace
  - deployment templates
  - deploy from source repo
  - optional separation of "control repo" and "gitops repo"
    - this lays the groundwork for the AD requirement to have multiple GitOps repos

### PR DESCRIPTION
**Pull request by bartr**
## Refactor

- support coral directory structure
- add auto Flux kustomization per workspace
- add support for coral templates
- change boa files to yaml
- change config file to yaml

## Type of PR

- [x] Documentation changes
- [x] Code changes
- [x] Test changes
- [ ] CI-CD changes
- [ ] GitHub Template changes

## Issues Closed or Referenced

- Closes https://github.com/retaildevcrews/retail-edge/issues/227
- Closes #369 
- Closes #421

### CODE REVIEW

File: bin/.coral/commands/gitops/delete [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+if [ "$CORAL_GITOPS_REPO" = "" ]
+then
+    echo "export CORAL_GITOPS_REPO=yourGitopsRepo first"
+    exit 0
+fi
+
+cd /workspaces || exit
+
+# delete the repos
+gh repo delete "$CORAL_CONTROL_REPO" --confirm
```
Comment (DanMass) [inline bin/.coral/commands/gitops/delete:18]: `--confirm` seems aggressive here. Does this operation really need to be this easy?

File: bin/.coral/coral-targets-add-completion [hunk 2]
```diff
@@ -0,0 +1,13 @@
+region:central	Central Region
```
Comment (DanMass) [inline bin/.coral/coral-targets-add-completion:1]: See above.

File: bin/.coral/coral-check-app-completion [hunk 3]
```diff
@@ -0,0 +1,12 @@
+heartbeat	Heartbeat app
+imdb	IMDb reference app
+jumpbox	Jump Box (bastion host)
+istio	Istio service mesh
+kustomize	Flux Kustomize operator
+coredns	K8s Core DNS
+metrics-server	Metrics Server
+testapp	Test App (sample app)
```
[the remark this answers is not in the document]
Comment (bartr) [reply bin/.coral/coral-check-app-completion:8]: this is just for demos and will be removed before merging

File: src/kic/boa/boa-yaml.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+			}
+		}
+	}
+
+	return nil
+}
+
+// load commands from a yaml file
+func readCommandFromYaml(fileName string) (subCmd, error) {
+	boaPath := GetBoaPath()
```
[the remark this answers is not in the document]
Comment (bartr) [reply src/kic/boa/boa-yaml.go:100]: fixed

File: src/kic/boa/boa-yaml.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+	if err != nil {
+		return nil, cfmt.ErrorE("unable to read file: ", file)
+	}
+
+	// map the yaml
+	result := make(map[string]interface{})
+
+	if err = yaml.Unmarshal(content, &result); err != nil {
+		return nil, cfmt.ErrorE("failed to unmarshal yaml")
```
[the remark this answers is not in the document]
Comment (bartr) [reply src/kic/boa/boa-yaml.go:74]: fixed

File: src/kic/boa/boa-yaml.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+	data, err := os.ReadFile(filepath.Join(boaPath, fileName))
+	if err != nil {
+		return subCmd{}, cfmt.ErrorE("unable to read file: ", filepath.Join(boaPath, fileName))
+	}
+
+	// unmarhal the yaml into the structure
+	root := subCmd{}
+	err = yaml.Unmarshal([]byte(data), &root)
+	if err != nil {
+		return subCmd{}, cfmt.ErrorE("invalid yaml in: ", filepath.Join(boaPath, fileName))
```
[the remark this answers is not in the document]
Comment (bartr) [reply src/kic/boa/boa-yaml.go:112]: same as above

File: src/kic/boa/boa-yaml.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+// commands are in the .binName directory
+//    that is a subdirectory of where the bin file is located
+// example: $HOME/bin/.kic
+func LoadCommands(appRootCmd *cobra.Command) {
+	// load boaPath/*.yaml
+	if files, err := ioutil.ReadDir(GetBoaPath()); err == nil {
+		for _, f := range files {
+			if !f.IsDir() && strings.HasSuffix(strings.ToLower(f.Name()), ".yaml") {
+				// read yaml
+				if cmd, err := readCommandFromYaml(f.Name()); err == nil {
```
[the remark this answers is not in the document]
Comment (bartr) [reply src/kic/boa/boa-yaml.go:47]: fixed

File: src/kic/boa/boa-yaml.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+	Path  string
+}
+
+// load commands from *.yaml
+// commands are in the .binName directory
+//    that is a subdirectory of where the bin file is located
+// example: $HOME/bin/.kic
+func LoadCommands(appRootCmd *cobra.Command) {
+	// load boaPath/*.yaml
+	if files, err := ioutil.ReadDir(GetBoaPath()); err == nil {
```
[the remark this answers is not in the document]
Comment (bartr) [reply src/kic/boa/boa-yaml.go:43]: fixed

[the remark this answers is not in the document]
Comment (bartr) [reply src/kic/boa/boa-yaml.go:43]: added error if cannot read dir

File: bin/.coral/commands/gitops/delete (hunk 1 above)
Comment (bartr) [reply bin/.coral/commands/gitops/delete:18]: this is demo code that will be removed before merge

File: src/kic/boa/boa-yaml.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+)
+
+type subCmd struct {
+	Name         string
+	Short        string
+	Long         string
+	Path         string
+	Hidden       bool
+	Commands     []subCmd
+	BoaCommands  []boaCmd `yaml:"boaCommands"`
```
[the remark this answers is not in the document]
Comment (bartr) [reply src/kic/boa/boa-yaml.go:25]: Good question. I thought about that and felt this approach gave more type safety and better organization to the yaml. 

In practice, most commands are going to use one type.

Here's an example where they are mixed: https://github.com/retaildevcrews/akdc/blob/reef/src/kic/.flt/root.yaml

I can add a "type" field and put all under one yaml heading.

I'm OK with either approach - we should make the change now so we don't have backwards compat issues. I'll send out an email to get some feedback.

**Review by aflinchb — APPROVED** [review]
Approved (pending design thread and removing coral pieces)

File: src/kic/boa/boa-yaml.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (bartr) [reply src/kic/boa/boa-yaml.go]: added issue to board for further discussion

**Pull request merged**